rcm_nexus/npm.py

```diff
@@ -0,0 +1,214 @@
+    :type message: str
+    """
+    sys.stderr.write(message)
+    exit(1)
+
+
+def push(nexus_config, repo, repo_type, product, debug=False):
+    """Publishes given repository
+    :param nexus_config: loaded configuration file
+    :type nexus_config: config.NexusConfig
+    :param repo: repository directory
+    :type repo: str
+    :param repo_type: type of the npm repository
+    :type repo_type: NpmArchiveType
+    :param product: product
+    :type product: str
+    :param debug: optional extended output
+    :type debug: bool
+    """
+
+    if not nexus_config.username or not nexus_config.password:
+        die('Missing user credentials! Please, supply user name and password!')
+
+    if not _npm_repository(nexus_config, product):
+        die(product + ' is not an npm product!')
+
+    if not nexus_config.get_npm_repository(product):
+        die('Deployment data for product {} were not found in the common configuration file!'.format(product))
+
+    expanded_repo = os.path.expanduser(repo)
+
+    tempdir = tempfile.mkdtemp()
+    try:
+        if repo_type == NpmArchiveType.DIRECTORY:
+            if expanded_repo[-1] != '/':
+                expanded_repo += '/'
+
+            if debug:
+                print("Temporary directory: "+tempdir)
+                print("Repository: "+expanded_repo)
+
+            copy_command = ('cp', '-rf', expanded_repo, tempdir)
+            subprocess.check_output(copy_command)
+
+            _publish_directory(nexus_config, tempdir, product, debug)
+
+        elif repo_type == NpmArchiveType.ZIP_FILE:
+            if debug:
+                print("Temporary directory: "+tempdir)
+                print("Repository: "+expanded_repo)
+
+            with zipfile.ZipFile(expanded_repo) as zz:
+                zz.extractall(tempdir)
+                _publish_directory(nexus_config, tempdir, product, debug)
+        elif repo_type == NpmArchiveType.TAR_FILE:
+            # create np
+            _create_npmrc_file(nexus_config, tempdir, product)
+
+            try:
+                shutil.copy2(expanded_repo, tempdir)
+            except IOError:
+                shutil.rmtree(tempdir)
+                die("Unable to copy {} to {}!".format(expanded_repo, tempdir))
+
+            package = os.path.join(tempdir, os.path.split(expanded_repo)[-1])
+
+            if debug:
+                print("Temporary directory: "+tempdir)
+                print("Repository: "+expanded_repo)
+                print("Target repository: "+package)
+
+            # publish the tarball using npm
+            command = ["npm", "publish", package]
+            if debug:
+                command.extend(["--verbose"])
+                print("Username: " + nexus_config.username)
+                if compare_digest(nexus_config.password, DEFAULT_PASSWORD):
+                    print("Password: not default password provided")
+                else:
+                    print("Password: default password used.")
```

Comment:
I'm not sure how this password handling should work. Instead of accessing `nexus_config.password`, you should call `get_password` method and that will handle any `@oracle` string that may be there.

I don't see the username or passward passed to the underlying `npm` command though. How does it authenticate against the remote service?

Reply:
we'll need to dig up an npm for RHEL-6 as well, since this runs on rcm-dev. There's a nodejs under eng-rhel-6, but not an npm.

Reply:
This is what I came here to find out. The tooling support will be challenging on the existing hosts. I wonder if we can run some node equivalent to virtualenv?